myia/parser.py

```diff
@@ -0,0 +1,958 @@
+        defaults_list = []
+
+        # We don't handle this for now.
+        assert args.posonlyargs == []
+
+        for arg, dflt in zip(pargs + kwargs, defaults + kwdefaults):
+            with debug_inherit(name=arg.arg, location=self.make_location(arg)):
+                param_node = Parameter(function_block.graph, name=arg.arg)
+
+            if arg.annotation:
+                param_node.add_annotation(self._eval_ast_node(arg.annotation))
+
+            function_block.graph.parameters.append(param_node)
+            function_block.write(arg.arg, param_node)
+            if dflt:
+                # TODO If there are no parents (the initial function),
+                # then evaluate in the global env
+                if block is None:
+                    raise MyiaSyntaxError(
+                        "default value on the entry function",
+                        self.make_location(arg),
+                    )
+                # XXX: This might not work correctly with our framework,
+                # but we need to evaluate the default arguments in the parent
+                # context for proper name resolution.
+                dflt_node = self.process_node(block, dflt)
+                defaults_name.append(arg.arg)
+                defaults_list.append(dflt_node)
+
+        if args.vararg:
+            arg = args.vararg
+            with debug_inherit(name=arg.arg, location=self.make_location(arg)):
+                vararg_node = Parameter(function_block.graph, name=arg.arg)
+            function_block.graph.parameters.append(vararg_node)
+            function_block.write(arg.arg, vararg_node)
+        else:
+            vararg_node = None
+
+        if args.kwarg:
+            arg = args.kwarg
+            with debug_inherit(name=arg.arg, location=self.make_location(arg)):
+                kwarg_node = Parameter(function_block.graph, name=arg.arg)
+            function_block.graph.parameters.append(kwarg_node)
+            function_block.write(arg.arg, kwarg_node)
+        else:
+            kwarg_node = None
+
+        function_block.graph.varargs = vararg_node
+        function_block.graph.kwargs = kwarg_node
+        function_block.graph.defaults = dict(zip(defaults_name, defaults_list))
+        function_block.graph.kwonly = len(args.kwonlyargs)
+
+    def _create_function(self, block, node):
+        with debug_inherit(name=node.name, location=self.make_location(node)):
+            function = Function(
+                parent=block,
+                flags=self.recflags,
+            )
+        function_block = function.initial_block
+
+        self._process_args(block, function_block, node.args)
+
+        after_block = self.process_statements(function_block, node.body)
+
+        if after_block.graph.return_ is None:
+            after_block.returns(Constant(None))
+
+        if node.returns:
+            function_block.graph.return_.add_annotation(
+                self._eval_ast_node(node.returns)
+            )
```

Comment:
Do we want to annotate `return_` or `output`? The latter might be more accurate if we decide to add more return values to `return_` like the universe etc. but I don't know.

Reply:
It is technically more accurate here since the annotation is on the function, not the output expression.

Also, the extra return values should be on different edges so we can separate them.